oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/NodeImpl.java

```diff
@@ -1015,6 +1015,9 @@ public void checkPreconditions() throws RepositoryException {
                     throw new VersionException(format(
                             "Cannot add mixin type. Node [%s] is checked in.", getNodePath()));
                 }
+                // OAK-10334: adding mixin requires permission to read existing mixin types
+                PropertyState prop = PropertyStates.createProperty(JCR_MIXINTYPES, singleton(oakTypeName), NAMES);
+                sessionContext.getAccessManager().checkPermissions(dlg.getTree(), prop, Permissions.READ_PROPERTY);
```

Comment:
hi @mreutegg , i wonder if it wouldn't be better to also adopt a similar approach as we use in {{NodeImpl.getPrimaryType}} and {{getMixinTypeNames}}.... i.e. making sure we don't overwrite the type in the tree-util by potentially reading the value from the read-only tree... in particular since TreeUtil.addMixins is also called internally afaik in the oak code.... in other words: this problem may still occur in case the mixin is added directly in the oak api (and not through JCR api) as this happens in the implementation.

also: if we want to keep the permission check here, we should also adjust Node.canAddMixin

Reply:
@anchela,
>  i wonder if it wouldn't be better to also adopt a similar approach as we use in {{NodeImpl.getPrimaryType}} and {{getMixinTypeNames}}.... i.e. making sure we don't overwrite the type in the tree-util by potentially reading the value from the read-only tree

Are you referring to OAK-2441, so basically bypassing the read restriction by force-reading through a `getReadOnlyTree`?

> also: if we want to keep the permission check here, we should also adjust Node.canAddMixin

+1 - except I guess if we went the `getReadOnlyTree` way, then this wouldn't be needed

Reply:
@stefan-egli , what i meant was: if the mixin property is not readable to the current session we could obtain the existing values from the read-only-tree. this would prevent the overwrite issue.

> if we went the getReadOnlyTree way, then this wouldn't be needed

not sure i get that.... but what i intended to suggest is to also keep Node.canAddMixin in mind and make sure the 2 methods are consistent. hope that helps.

Reply:
>  i wonder if it wouldn't be better to also adopt a similar approach as we use in `NodeImpl.getPrimaryType` and `getMixinTypeNames`.... i.e. making sure we don't overwrite the type in the tree-util by potentially reading the value from the read-only tree...

That's a good point and I think would be more consistent. Oak and Jackrabbit allow you to get primary and mixin type information even if the session does not have permission to read properties where those values are stored in. It would be strange if adding a mixin fails in this case.

I'll update the PR accordingly.

Reply:
Done.

@anchela, please have a look at the changes again. Thanks.

Reply:
It appears Angela is away. @stefan-egli, can you please review when you get a chance?